File: 006.zigzag-conversion/zigzag-conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        lenght = len(s)
        step = 2 * (numRows - 1)
        result = s[::step]
        for i in range(1, numRows - 1):
            index = i
            while index < lenght:
                result += s[index]
                index += step - i*2
                if index < lenght:
                    result += s[index]
                    index += i * 2
        result += (s[numRows-1::step])

        return result


if __name__ == '__main__':
    s = Solution()
    convert = s.convert

    print(convert("PAYPALISHIRING", 3) == "PAHNAPLSIIGYIR")
    print(convert(s="LEETCODEISHIRING", numRows=3) == "LCIRETOESIIGEDHN")
    print(convert(s="LEETCODEISHIRING", numRows=4) == "LDREOEIIECIHNTSG")

File: 006.zigzag-conversion/zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  const step = 2 * (numRows - 1)
  const len = s.length
  let result = ''

  for (let i = 0; i < len; i += step) {
    result += s[i]
  }

  for (let i = 1; i < numRows - 1; i += 1) {
    let index = i
    // console.log(len)
    while (index < len) {
      // console.log(s[index])
      result += s[index]
      index += step - 2 * i
      if (index < len) {
        result += s[index]
        index += 2 * i
      }
    }
  }

  for (let i = numRows - 1; i < len; i += step) {
    result += s[i]
  }

  return result
};

console.log(convert("PAYPALISHIRING", 3) === "PAHNAPLSIIGYIR")
console.log(convert(s = "LEETCODEISHIRING", numRows = 3) == "LCIRETOESIIGEDHN")
console.log(convert(s = "LEETCODEISHIRING", numRows = 4) == "LDREOEIIECIHNTSG")
